File: src/main/html5/js/thingcontainer.js

function ThingContainer(nthings) {
	this.__setup();
}

ThingContainer.prototype.__setup = function(nthings, layer) {
	this.n = nthings;
	this.things = new Array(this.n);
	this.layers = new Array(layer);
};

ThingContainer.prototype.addThings = function(objs, p, layer) {
	this.pos = p;
	var ninlayer = 0;
	this.layers[layer] = p;
	for ( var i = 0; i < this.n; i++) {
		if (this.things[i].getLayer() == layer) {
			// console.log("adding: "+this.things[i].getName()+" at: "+p);
			objs[p++] = this.things[i];
			ninlayer++;
		}
	}
	return ninlayer;
};

ThingContainer.prototype.removeThings = function(objs) {
	var lastlayer = -1;
	var thingsinlayer = 0;
	for ( var i = 0; i < this.n; i++) {
		var layer = this.things[i].getLayer();
		var p = this.layers[layer];

		if (lastlayer != layer)
			thingsinlayer = 0;
		else
			thingsinlayer++;

		objs[p + thingsinlayer] = null;
	}
};

ThingContainer.prototype.translate = function(x, y, z) {
	for ( var i = 0; i < this.n; i++) {
		this.things[i].translate(x, y, z);
	}
};

ThingContainer.prototype.rotate = function(deg) {
	for ( var i = 0; i < this.n; i++) {
		this.things[i].rotate(deg);
	}
};

ThingContainer.prototype.scale = function(sx, sy) {
	for ( var i = 0; i < this.n; i++) {
		this.things[i].scale(sx, sy);
	}
};

ThingContainer.prototype.translateRoot = function(x, y, z) {
	for ( var i = 0; i < this.n; i++) {
		this.things[i].translateRoot(x, y, z);
	}
};

ThingContainer.prototype.rotateRoot = function(deg) {
	for ( var i = 0; i < this.n; i++) {
		this.things[i].rotateRoot(deg);
	}
};

ThingContainer.prototype.scaleRoot = function(sx, sy) {
	for ( var i = 0; i < this.n; i++) {
		this.things[i].scaleRoot(sx, sy);
	}
};

ThingContainer.prototype.getType = function() {
	return THINGCONTAINER;
};

File: src/main/java/de/digitalemil/eagle/ThingContainer.java
package de.digitalemil.eagle;

import de.digitalemil.tocplusplus.MethodDefinitionChangerAnnotation;

public class ThingContainer {
	protected int n;
	protected Thing things[];
	protected int layers[];
	protected int pos;

	public ThingContainer() {
	}

	@MethodDefinitionChangerAnnotation({ "BY", "new Thing", "new Thing*" })
	public void init(int nthings) {
		n = nthings;
		things = new Thing[n];
		layers = new int[1000];
	}

	@MethodDefinitionChangerAnnotation({"BY", "things[i]=null", "delete things[i]", "BY", "things=null", "delete [ ] things", "BY", "layers=null", "delete [ ] layers"})
	protected void finalize() throws Throwable {
		for (int i = 0; i < n; i++) {
			things[i] = null;
		}
		things = null;
		layers = null;
	}

	public int addThings(Thing[] in, int p, int layer) {
		int ninlayer = 0;
		for (int i = 0; i < n; i++) {
			if (things[i].getLayer() == layer) {
				// System.out.println("adding: "+things[i].getName()+" at: "+p);
				in[p++] = things[i];
				layers[layer] = p;
				ninlayer++;
			}
		}
		return ninlayer;
	}

	public void removeThings(Thing[] objs) {
		int lastlayer = -1;
		int thingsinlayer = 0;
		for (int i = 0; i < n; i++) {
			int layer = things[i].getLayer();
			int p = layers[layer];

			if (lastlayer != layer)
				thingsinlayer = 0;
			else
				thingsinlayer++;

			objs[p + thingsinlayer] = null;
		}
	}

	public void beginTX() {
		for (int i = 0; i < n; i++) {
			if (things[i] != null)
				things[i].beginTX();
		}
	}

	public void commitTX() {
		for (int i = 0; i < n; i++) {
			if (things[i] != null)
				things[i].commitTX();
		}
	}

	public void rollbackTX() {
		for (int i = 0; i < n; i++) {
			if (things[i] != null)
				things[i].rollbackTX();
		}
	}

	public void translate(float ix, float iy, float iz) {
		for (int i = 0; i < n; i++) {
			if (things[i] != null)
				things[i].translate(ix, iy, iz);
		}
	}

	public void rotate(float deg) {
		for (int i = 0; i < n; i++) {
			if (things[i] != null)
				things[i].rotate(deg);
		}
	}

	public void scale(float sx, float sy) {
		for (int i = 0; i < n; i++) {
			if (things[i] != null)
				things[i].scale(sx, sy);
		}
	}

	public void translateRoot(float x, float y, float z) {
		for (int i = 0; i < n; i++) {
			if (things[i] != null)
				things[i].translateRoot(x, y, z);
		}
	}

	public void rotateRoot(float deg) {
		for (int i = 0; i < n; i++) {
			if (things[i] != null)
				things[i].rotateRoot(deg);
		}
	}

	public void scaleRoot(float sx, float sy) {
		for (int i = 0; i < n; i++) {
			if (things[i] != null)
				things[i].scaleRoot(sx, sy);
		}
	}

	public int getType() {
		return Types.THINGCONTAINER;
	}

}
